docs/integrations/step-patterns.md

```diff
@@ -0,0 +1,326 @@
+build relationships to downstream entities such as `User`s or `Group`s, so we
+typically store this `Account` entity in our job state for easy retrieval later.
+
+The archetype for the **Singleton** pattern:
+
+```typescript
+const fetchAccountStep: Step = {
+  id: 'fetch-account',
+  name: 'Fetch Account',
+  entities: [{ _type: 'acme_account', ... }],
+  relationships: [],
+  dependsOn: [],
+  executionHandler: (context) => {
+    const { jobState, instance } = context;
+    const client = new AcmeClient({ config: instance.config });
+
+    const account = await client.getAccount();
+
+    const accountEntity = createAccountEntity(account);
+    await jobState.addEntity(accountEntity);
+    await jobState.setData(accountEntity, 'ACCOUNT_ENTITY');
+  },
+}
+```
+
+## "Fetch Entities" step pattern
+
+Endpoints that return a list of resources from a path such as `/users`,
+`/groups`, or `/devices`, without path parameters, should typically be ingested
+using the **Fetch Entities** pattern. Sometimes, but not always, these basic
+list resources have an implicit relationship back to the `Account` entity.
+
+The archetype for the **Fetch Entities** pattern:
+
+```typescript
+const fetchUsersStep: Step = {
+  id: 'fetch-users',
+  name: 'Fetch Users',
+  entities: [{ _type: 'acme_user', ... }],
+  relationships: [{ _type: 'acme_account_has_user', ... }],
+  dependsOn: ['fetch-account'],
+  executionHandler: (context) => {
+    const { jobState, instance } = context;
+    const accountEntity = await jobState.getData('ACCOUNT_ENTITY');
+    const client = new AcmeClient({ config: instance.config });
+
+    await client.iterateUsers(
+      async (user) => {
+        const userEntity = createUserEntity(user);
+        await jobState.addEntity(userEntity);
+
+        const accountUserRelationship = createDirectRelationship({
+          from: accountEntity,
+          _class: RelationshipClass.HAS,
+          to: userEntity,
+        });
+        await jobState.addRelationship(accountUserRelationship);
+      }
+    );
+  },
+}
+```
+
+## "Fetch Relationships" step pattern
+
+Endpoints that list resources using path parameters, such as
+`/groups/{groupId}/users`, typically indicate either the **Fetch Relationships**
+pattern or the **Fetch Child Entities** pattern, described below. Target
+resources (in this case, `acme_user`) otherwise listed from a top-level **list**
+endpoint (like `/users`), signal that JupiterOne entities will be created
+elsewhere using a **Fetch Entities** pattern and that this endpoint should be
+handled with a **Fetch Relationships** pattern.
+
+This type of step ingests no new entities, and instead builds relationships
+between existing entities. This pattern uses the `jobState.iterateEntities` and
+`jobState.findEntity` methods to reference entities that have already been added
+to the job state in previous steps. It's important that both the `fetch-groups`
```

Comment:
Let's chat more about this in the future. I think verifying that the `_key` exists in the job state before building the relationship is very useful. Maybe it's better to use `jobState.hasKey()` rather than `jobState.findEntity` in that case.